numpy/_core/src/common/npy_hashtable.c

```diff
@@ -210,10 +210,13 @@ _resize_if_necessary(PyArrayIdentityHash *tb)
  * @param value Normally a Python object, no reference counting is done.
```

Comment:
Wouldn't it be better to do reference counting? Then, the caller knows it can safely DECREF the value any time the function returns without error.

Reply:
I agree it would be better to do reference counting, but it's a preexisting issue here and I was hoping I could get away without fixing it :)

I looked a bit at this just now and I think to fix the reference counting I'd need to also make `resolve_implementation_info` in `dispatching.c` return a new reference instead of a borrowed reference. Maybe that's not so bad though because there's only one call site.

Still, I'd prefer to work on this in a followup if we can.

Reply:
I agree, it would be a follow up.  One idea behind no refcounting here, was that it would be nice if neither the keys (dtypes) nor the values (arraymethods) had to be strictly immortal.

But, it turns out that for most cases, that isn't just not that important, but rather impossible because Python lacks [emphemerons](https://en.wikipedia.org/wiki/Ephemeron) (a concept necessary to clean up ufunc signatures).